Warnings -Wsign-compare inside an assert when building with Clang 11.1.0 on Windows

The exact problem happens when assert() macro is active; the code location in release 0.8.0 is:

xxhsum.c:1540:38: warning: comparison of integers of different signs: 'const AlgoSelected' and 'unsigned long long' [-Wsign-compare]
assert(0 <= hashType && hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));

This is breaking build of projects using xxHash as in-source dependency, when -Werror flag is active.

Thanks for the info.  I've successfully reproduce this issue with `LLVM-11.1.0-win64.exe` on `Microsoft Windows [Version 10.0.19042.1165]`.

For now, I just put the procedure to help other contributors can follow this issue.

## Install LLVM-11.1.0-win64

```bat
ver
cd /d "%USERPROFILE%\Downloads"
curl -JLO https://github.com/llvm/llvm-project/releases/download/llvmorg-11.1.0/LLVM-11.1.0-win64.exe
start .\LLVM-11.1.0-win64.exe
```

## v0.8.0

```bat
cd /d "%USERPROFILE%\Documents"
git clone https://github.com/Cyan4973/xxHash.git xxHash-issue-588-v0.8.0
cd xxHash-issue-588-v0.8.0
git checkout v0.8.0
git log -1

set "clang=C:\Program Files\LLVM\bin\clang.exe"
"%clang%" -Wall -Wextra -Werror xxhash.c xxhsum.c
```

```txt
In file included from xxhash.c:43:
./xxhash.h:3503:1: error: unused function 'XXH3_accumulate_512_scalar' [-Werror,-Wunused-function]
XXH3_accumulate_512_scalar(void* XXH_RESTRICT acc,
^
./xxhash.h:3521:1: error: unused function 'XXH3_scrambleAcc_scalar' [-Werror,-Wunused-function]
XXH3_scrambleAcc_scalar(void* XXH_RESTRICT acc, const void* XXH_RESTRICT secret)
^
./xxhash.h:3538:1: error: unused function 'XXH3_initCustomSecret_scalar' [-Werror,-Wunused-function]
XXH3_initCustomSecret_scalar(void* XXH_RESTRICT customSecret, xxh_u64 seed64)
^
3 errors generated.
xxhsum.c:1540:38: error: comparison of integers of different signs: 'const AlgoSelected' and 'unsigned long long'
      [-Werror,-Wsign-compare]
    assert(0 <= hashType && hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\Include\10.0.20348.0\ucrt\assert.h:40:17: note: expanded from macro 'assert'
            (!!(expression)) ||                                                              \
                ^~~~~~~~~~
xxhsum.c:1562:38: error: comparison of integers of different signs: 'const AlgoSelected' and 'unsigned long long'
      [-Werror,-Wsign-compare]
    assert(0 <= hashType && hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\Include\10.0.20348.0\ucrt\assert.h:40:17: note: expanded from macro 'assert'
            (!!(expression)) ||                                                              \
                ^~~~~~~~~~
2 errors generated.
```


## dev

```bat
cd /d "%USERPROFILE%\Documents"
git clone https://github.com/Cyan4973/xxHash.git xxHash-issue-588-dev
cd xxHash-issue-588-dev
git log -1

set "clang=C:\Program Files\LLVM\bin\clang.exe"
"%clang%" -Wall -Wextra -Werror xxhash.c cli/xxhsum.c cli/xsum_os_specific.c cli/xsum_output.c cli/xsum_sanity_check.c
```

```txt
In file included from xxhash.c:43:
./xxhash.h:4160:1: error: unused function 'XXH3_accumulate_512_scalar' [-Werror,-Wunused-function]
XXH3_accumulate_512_scalar(void* XXH_RESTRICT acc,
^
./xxhash.h:4178:1: error: unused function 'XXH3_scrambleAcc_scalar' [-Werror,-Wunused-function]
XXH3_scrambleAcc_scalar(void* XXH_RESTRICT acc, const void* XXH_RESTRICT secret)
^
./xxhash.h:4195:1: error: unused function 'XXH3_initCustomSecret_scalar' [-Werror,-Wunused-function]
XXH3_initCustomSecret_scalar(void* XXH_RESTRICT customSecret, xxh_u64 seed64)
^
3 errors generated.
cli/xxhsum.c:592:38: error: comparison of integers of different signs: 'const AlgoSelected' and 'unsigned long long' [-Werror,-Wsign-compare]
    assert(0 <= hashType && hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\Include\10.0.20348.0\ucrt\assert.h:40:17: note: expanded from macro 'assert'
            (!!(expression)) ||                                                              \
                ^~~~~~~~~~
cli/xxhsum.c:614:38: error: comparison of integers of different signs: 'const AlgoSelected' and 'unsigned long long' [-Werror,-Wsign-compare]
    assert(0 <= hashType && hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\Include\10.0.20348.0\ucrt\assert.h:40:17: note: expanded from macro 'assert'
            (!!(expression)) ||                                                              \
                ^~~~~~~~~~
2 errors generated.
```

--- a/cli/xxhsum.c
+++ b/cli/xxhsum.c
@@ -106,11 +106,6 @@ static AlgoSelected g_defaultAlgo = algo_xxh64;    /* required within main() & X
 
 
 /* ************************************
- *  Display macros
- **************************************/
-
-
-/* ************************************
  *  Local variables
  **************************************/
 static XSUM_U32 g_nbIterations = NBLOOPS;
@@ -589,7 +584,7 @@ static void XSUM_printLine_BSD_internal(const char* filename,
                                         const char* algoString[],
                                         XSUM_displayHash_f f_displayHash)
 {
-    assert(0 <= hashType && hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));
+    assert(0 <= hashType && (size_t)hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));
     {   const char* const typeString = algoString[hashType];
         const size_t hashLength = XSUM_algoLength[hashType];
         XSUM_output("%s (%s) = ", typeString, filename);
@@ -611,7 +606,7 @@ static void XSUM_printLine_GNU_internal(const char* filename,
                                const void* canonicalHash, const AlgoSelected hashType,
                                XSUM_displayHash_f f_displayHash)
 {
-    assert(0 <= hashType && hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));
+    assert(0 <= hashType && (size_t)hashType <= XSUM_TABLE_ELT_SIZE(XSUM_algoName));
     {   const size_t hashLength = XSUM_algoLength[hashType];
         f_displayHash(canonicalHash, hashLength);
         XSUM_output("  %s\n", filename);
@@ -1415,7 +1410,7 @@ XSUM_API int XSUM_main(int argc, char* argv[])
             /* Display version */
             case 'V':
                 XSUM_log(FULL_WELCOME_MESSAGE(exename));
-                XSUM_sanityCheck(); 
+                XSUM_sanityCheck();
                 return 0;
 
             /* Display help on XSUM_usage */

--- a/cmake_unofficial/CMakeLists.txt
+++ b/cmake_unofficial/CMakeLists.txt
@@ -42,6 +42,10 @@ endif()
 if(NOT CMAKE_CONFIGURATION_TYPES)
   message(STATUS "xxHash build type: ${CMAKE_BUILD_TYPE}")
 endif()
+# Enable assert() statements in debug builds
+if("${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
+  add_compile_definitions(XXH_DEBUGLEVEL=1)
+endif()
 
 option(BUILD_SHARED_LIBS "Build shared library" ON)
 option(XXHASH_BUILD_XXHSUM "Build the xxhsum binary" ON)

